---
title: "DESeq_Visualization"
author: "clagrabel"
date: "2024-03-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading the necessary packages

```{r}
#BiocManager::install("apeglm", force=TRUE)
library(apeglm)

pacman::p_load(tidyverse, DESeq2, ggVennDiagram, UpSetR, plotly, ggrepel, scico)
```

# Loading the data

```{r}
# DEG object
dds<-readRDS("./results/deseq2_dds.rds")
dds_test<-readRDS("./results/deseq2_dds_test.rds")

# The list of DEG results
res<-readRDS("./results/deseq2_results_populations.rds")
test<-readRDS("./results/deseq2_results_test.rds")
summary(res)
summary(test)

```

# Venn Diagrams

Extract DEGs lists in order to visualize overlapping regulated genes in the comparisons of interest.


```{r}
# List of DEGs

# Creating a function to extract the lists of genes repeteadly: all DEGs, up-regulated DEGs and down-regulated DEGs.

extract_degs<-function(x) {
  return(
    x %>%
      as_tibble(rownames = "gene") %>%
      filter(padj<0.05) %>%
      pull(gene)
  )
}

extract_up<-function(x) {
  return(
    x %>%
      as_tibble(rownames = "gene") %>%
      filter(padj<0.05) %>%
      filter(log2FoldChange>0) %>%
      pull(gene)
  )
}

extract_down<-function(x) {
  return(
    x %>%
      as_tibble(rownames = "gene") %>%
      filter(padj<0.05) %>%
      filter(log2FoldChange<0) %>%
      pull(gene)
  )
}

# Extracting all differentially expressed genes from all the comparisons stored in the list of DESeqResults.

all_degs<-lapply(res, FUN=extract_degs)
str(all_degs)

up_degs<-lapply(res, FUN=extract_up)
down_degs<-lapply(res, FUN=extract_down)

```


```{r eval=FALSE, include=FALSE}
#My way of doing venn diagrams

library(ggvenn)

venn_data <- list(nombre1=vector1,
                  nombre2=vector2,
                  nombre3=vector3)

venn_palette <- c("#e35023","#148733","#058dfc")

ggvenn(venn_data, columns = c("nombre1","nombre2","nombre3"),stroke_size = 0.3,
       fill_color = venn_palette, stroke_color="white", show_percentage=F,
       fill_alpha=0.6, set_name_color = venn_palette, text_size=6)
```


# Venn Diagrams

## Regulated genes in High and Low water in both Regions

It is interesting to check both up and down-regulated genes in each region in the different treatment conditions (high and low water).
This Venn Diagram would bring light over genes that are regulated in the Southern region which are not in the Central region and viceversa.


```{r}
all_degs[c("lS_hS","lC_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") + 
  ggtitle("DEGs in Southern Spain vs. DEGs in Central Spain")
```


## Regulated genes for different populations belonging to the same region

When looking at these diagrams, we expect to obtain the genes that are overlapping or not in the different populations belonging to the same region.

### Central Region

```{r}
all_degs[c("hBui_lBui","hCan_lCan", "hTur_lTur")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") + 
  ggtitle("DEGs in the populations from Central Spain")
```


### Southern Region

```{r}
all_degs[c("hEsp_lEsp","hLla_lLla", "hJab_lJab")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") + 
  ggtitle("DEGs in the populations from Southern Spain")
```


## Up and Down-regulated genes


```{r}
up_degs[c("lS_hS","lC_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") +
  ggtitle("Up-regulated genes South and Central")
```

```{r}
down_degs[c("lS_hS","lC_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") +
  ggtitle("Down-regulated genes South and Central")
```



# Test for all populations except Llano

## Regulated genes in High and Low water in both Regions

It is interesting to check both up and down-regulated genes in each region in the different treatment conditions (high and low water).
This Venn Diagram would bring light over genes that are regulated in the Southern region which are not in the Central region and viceversa.


```{r}
all_degs_test<-lapply(test, FUN=extract_degs)
str(all_degs_test)

up_degs_test<-lapply(test, FUN=extract_up)
down_degs_test<-lapply(test, FUN=extract_down)
```


```{r}
all_degs_test[c("test_lS_hS","test_lC_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") + 
  ggtitle("DEGs regarding water level in Southern Spain vs. DEGs in Central Spain")
```

Regarding water level response, there are 53 DEGs in Central Spain and 27 in Southern Spain. There are no common differentially expressed genes between Southern and Central regions.


## Up and Down-regulated genes


```{r}
up_degs_test[c("test_lS_hS","test_lC_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") +
  ggtitle("Up-regulated genes regarding water level South and Central")
```

Out of the DEGs, there are 48 genes that are up-regulated in Central Spain versus 11 in the South.


```{r}
down_degs_test[c("test_lS_hS","test_lC_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") +
  ggtitle("Down-regulated genes regarding water level South and Central")
```

However, the amount of down-regulated genes in Southern Spain is higher than in Central Spain.

### Regulated genes between regions for the same water level

```{r}
all_degs_test[c("test_lS_lC","test_hS_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow") + 
  ggtitle("DEGs in Southern Spain vs. DEGs in Central Spain")
```

Still, the larger effect is due to the region and not the water level. There is a higher amount of DEGs in the South compared to the Center, and there is a large number of common shared DEGs between regions.


## Venn Diagram all comparisons

From this diagram we can extract information about common DEGs for different comparisons.

```{r}
str(all_degs_test)

all_degs_test[c("test_lS_hS","test_lC_hC", "test_lS_lC", "test_hS_hC")] %>%
  ggVennDiagram(edge_size = 0) +
  scale_fill_scico(palette = "batlow")
```


```{r}
list_test<-all_degs_test[c("test_lS_hS","test_lC_hC", "test_lS_lC", "test_hS_hC")]

upset(fromList(list_test),
      number.angles = 0, point.size = 3, line.size = 1,
      sets.x.label = "Number of DEGs",
      set_size.show = TRUE,
      set_size.scale_max = max(sapply(list_test, length)), 
      text.scale = c(1.2, 1.2, 1.2, 1.2, 1.5, 1.5),
      order.by=c("degree","freq"))
```



# Functional Enrichment

We will study whether genes belonging to specific annotated gene sets, such as Gene Ontology (GO) terms, are enriched in the set of genes of interest, in this case the list of differentially expressed genes (DEGs). This analysis is commonly referred to as Over-Representation Analysis (ORA).

To conduct such an analysis, we need:

* A set of genes obtained from experimental data that are biologically relevant for our research, in this case the list of DEGs. These genes must be annotated, meaning we need information about which known biological functions they are associated to. For this, we will make use of an annotation file generated through BLASTing _Pelobates cultripes_ transcripts against the Ensembl _Xenopus tropicalis_ proteome using diamond (v2.1.8).

* A 'background' set of genes, which includes all genes that could potentially have been differentially expressed. This background set serves as the reference against which the enrichment of specific gene sets is assessed.

* A curated database containing biologically relevant categories, such as GO terms or gene pathways, along with the genes associated with each category. 


## Loading the annotation

This annotation file contains all _P. cultripes_ transcripts by rows. As columns, we can find:
* The gene IDs for _P. cultripes_, foll.owed by the transcripts IDs and the peptides IDs (gene_id, transcript_id, peptide_id)
* The IDs  and descriptions of _X. tropicalis_ annotated proteome resulting from both nucleotide and peptide blasting against _P. cultripes_ transcripts (xenx_pep_id, xenx_gene_symbol, xenx_description, xenp_pep_id, xenp_gene_symbol, xenp_description).


```{r}
# The list of DEG results was already loaded

# The annotation file
xtrop<-read_csv("diamondblast109.csv")
xtrop
```

Since we will be repeateadly extracting the IDs from this annotation file for the multiple comparisons, we will make a function that extracts matching _X. tropicalis_ IDs.

We will first focus on the IDs resulting from BLASTP.

# DOUBT

Is it broader from the nucleotide blast? therefore I can get more annotated genes, but peptides are more precise.

# DOUBT

diamondblast109.csv

Is gprofiler already taking into account the multiple annotations separated by ;??

Possible ways of using this annotation file?
* Consider each annotation separately (what I thought, selecting only 1 annotation resulting from a blast).
* Combine annotations somehow, aggregation? 
  + for many genes there is only the xenx_pep_id and xenp_pep_id, for some there is the description too




```{r}
extract_xtr<-function(x) {
  return(
      xtrop %>%
        filter(gene_id %in% x) %>% #Filtering data frame (xtrop) to select rows where the gene_id column matches any of the IDs provided in the input list (x).
        pull(xenp_pep_id) %>% #Extracts the values from the xenp_pep_id column of the filtered data frame.
        unique()
  )
}

# Applying this function to a list of Pelobates IDs
xtr_deg<-lapply(all_degs_test, FUN=extract_xtr)
str(all_degs_test)
str(xtr_deg)
```

Very few genes were not found in the annotation.


# Make background

As the background, we will use the all the genes that could be annotated in the Pelobates genome. 
This way, we ensure that the background includes the genes that are potentially relevant to our study system while capturing the variability in gene expression across different conditions.

Since _X. tropicalis_ is a different species, using all genes annotated in the Xenopus proteome could introduce genes that are not present in Pelobates, or may not accurately represent the genes that are relevant to the specific conditions of this study (regions and water level).


```{r eval=FALSE, include=FALSE}

rownames_bg<-unique(c(rownames(test_waterLevel), rownames(test_region), rownames(test_interaction), rownames(test_lS_hS), rownames(test_lC_hC), rownames(test_hS_hC), rownames(test_lS_lC)))

head(rownames_xtr_bg)
length(rownames_xtr_bg)
class(rownames_xtr_bg)

xtr_background<-lapply(rownames_bg, FUN=extract_xtr)

head(xtr_background)
class(xtr_background)

all_degs_test$test_waterLevel
rownames(all_degs_test)
rownames(xtr_background)

str(all_degs_test)
str(xtr_background)
```

```{r}
all_xtr_bg<-unique(c(extract_xtr(rownames(test$test_lC_hC)), extract_xtr(rownames(test$test_lS_hS)), extract_xtr(rownames(test$test_lS_lC)), extract_xtr(rownames(test$test_hS_hC)), extract_xtr(rownames(test$test_waterLevel)), extract_xtr(rownames(test$test_region)), extract_xtr(rownames(test$test_interaction))))

class(all_xtr_bg)
```



## Gene Ontology Enrichment

We will use g:profiler tool, which is a web-based tool for functional enrichment analysis of gene lists. 

The associated R package for g:Profiler is gprofiler2, which is an API (Application Programming Interface). This API serves as a bridge that allows performing the actual analysis within the g:Profiler server, by using the R environment in RStudio.

g:Profiler is continuously being updated to match the updates in the associated Ensembl and Ensembl Genomes Data Bases. Because of this, it is important to specify which version of g:Profiler, and therefore Ensembl version, we would like to use. 
We will do this by looking at the archives on the g:Profiler homepage and setting the base url for the desired version.

Our annotations resulted from BLASTing _P. cultripes_ transcriptome against the 109 Ensembl release of the _X. tropicalis_ proteome, therefore we will use Ensembl 109, Ensembl Genomes 56 (database built on 2023-03-29).


```{r}
# Install and load the gprofiler2 package. This package provides functions to access the gprofiler API directly from R. 

# install.packages("gprofiler2")
library(gprofiler2)

# Setting the base url

set_base_url("https://biit.cs.ut.ee/gprofiler_archive3/e109_eg56_p17")
# xtr_background
# xtr_bg_vector<-unlist(xtr_background)
# xtr_bg_vector
# class(as.vector(xtr_background))

# Run the analysis
test_ora<-gost(multi_query = FALSE, # returns separate results tables for multiquery
              custom_bg = all_xtr_bg, # our background
              query=xtr_deg, # our list of gene sets
              organism="xtropicalis", # the organism our annotations belong to
              exclude_iea = FALSE, # include GO terms that were electronically assigned
              correction_method = "gSCS", # the recommended multiple testing correction.
              sources=c("GO:BP","GO:CC","GO:MF", "KEGG","REAC"), # the functional sets we are interested in 
              evcodes=FALSE, # evcodes TRUE needed for downstream analysis like enrichment maps in Cytoscape, but it takes longer.
              significant= FALSE) # return all terms, not just the significant ones


# The results are stored as a "results" dataframe 
head(test_ora$result)
dim(test_ora$result)
```


Let's look at the enrichment results in detail

```{r}
colnames(test_ora$result)
```

* **query**: This column represents the gene set or query submitted to g:Profiler for enrichment analysis. Each row corresponds to a different gene set or query.

* **significant**: This column indicates whether the enrichment analysis identified the term (functional category) as statistically significant. It contains a boolean value (TRUE or FALSE).

* **p_value**: This column contains the adjusted p-values. Each p-value indicates the statistical significance of enrichment of a given term in the gene set.

* **term_size**: This column represents the total number of genes in a specific term (functional category).

* **query_size**: This column represents the total number of genes in the submitted query or gene set.

* **intersection_size**: This column represents the number of genes that are shared between the submitted query and the term being analyzed. That is, how many genes from both term size and query size are overlapping. 

* **term_id**: This column contains the unique identifier (ID) associated with the term (functional category) being analyzed.

* **term_name: This column contains the name or description of the term (functional category) being analyzed.**

* source: This column indicates the data source or database from which the term originates and that we requested (Gene Ontology, KEGG, Reactome).

* precision: This column represents the proportion of the genes shared between the submitted query and the term being analyzed, relative to the total number of genes in the query. It measures the accuracy of the enrichment analysis. A higher precision value indicates that a larger proportion of the genes in the intersection are relevant to the term being analyzed, relative to the total number of genes in the query. In other words, it reflects how well the term represents the genes in the query.

* recall: This column represents the proportion of the genes in the intersection, the same as the precision column, but relative to the total number of genes in the term. It reflects how well the term represents the genes in the dataset or background.

* effective_domain_size: This column represents the effective domain size used in the enrichment analysis. It is a statistical parameter used to calculate the p-value.

* source_order: This column represents the order or rank of the source (data database) based on its importance or relevance in the enrichment analysis.

* parents: This column contains information about the parent terms or higher-level categories to which the analyzed term belongs. It may provide additional hierarchical context for the analyzed terms.

