---
title: "DEGs_Visualization_Annotations"
author: "clagrabel"
date: "2024-06-23"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# DEGs Visualization

### Loading the necessary packages

```{r}
pacman::p_load(tidyverse, DESeq2, ggVennDiagram, UpSetR, plotly, ggrepel, scico, apeglm)
```


## Loading the data

```{r}

# The list of DEG results
res<-readRDS("./results/deseq2_regions_results_local.rds")
summary(res)

```



## Loading the Annotations


This annotation file contains all _P. cultripes_ transcripts by rows. As columns, we can find:
* The gene IDs for _P. cultripes_, followed by the transcripts IDs and the peptides IDs (gene_id, transcript_id, peptide_id)
* The IDs  and descriptions of _X. tropicalis_ annotated proteome resulting from both nucleotide and peptide blasting against _P. cultripes_ transcripts (xenx_pep_id, xenx_gene_symbol, xenx_description, xenp_pep_id, xenp_gene_symbol, xenp_description).


```{r}
# The annotation file
xtrop<-read.csv("./xtr109/diamondblast109.csv", stringsAsFactors = FALSE)

```



## Extracting Significant DEGs

Extract DEGs lists in order to visualize overlapping regulated genes in the comparison of interest.

```{r}
# List of DEGs

# Creating a function to extract the lists of genes repeatedly: all DEGs, up-regulated DEGs and down-regulated DEGs.

extract_degs<-function(x) {
  return(
    x %>%
      as_tibble(rownames = "gene") %>%
      filter(padj<0.05) %>%
      pull(gene)
  )
}

extract_up<-function(x) {
  return(
    x %>%
      as_tibble(rownames = "gene") %>%
      filter(padj<0.05) %>%
      filter(log2FoldChange>0) %>%
      pull(gene)
  )
}

extract_down<-function(x) {
  return(
    x %>%
      as_tibble(rownames = "gene") %>%
      filter(padj<0.05) %>%
      filter(log2FoldChange<0) %>%
      pull(gene)
  )
}

# Extracting all differentially expressed genes from all the comparisons stored in the list of DESeqResults.

sig_degs<-lapply(res, FUN=extract_degs)
str(sig_degs)

up_degs<-lapply(res, FUN=extract_up)
down_degs<-lapply(res, FUN=extract_down)

```


## Venn Diagram


```{r}
library(ggvenn)

sig_degs_set <- list(Central=sig_degs$central,
                  South=sig_degs$south)

venn_palette <- c("dodgerblue","darkblue")

ggvenn(sig_degs_set, columns = c("Central","South"),stroke_size = 0.3,
       fill_color = venn_palette, stroke_color="black", show_percentage=F,
       fill_alpha=0.6, set_name_color = "black", text_size=6)

```


### Upset Plot


```{r}

# Plot Upset

upset(fromList(sig_degs),
      nsets = length(sig_degs),
      keep.order = T,
      nintersects = 100,
      number.angles = 0, point.size = 3, line.size = 1,
      sets.x.label = "Number of DEGs",
      set_size.show = TRUE,
      set_size.scale_max = max(sapply(sig_degs, length))+200, 
      text.scale = c(1.2, 1.2, 1.2, 1.2, 1.5, 1.5),
      sets.bar.color = c("dodgerblue","darkblue"),
      order.by=c("degree","freq"))

```



# Volcano plot

Showing the log fold change plotted against the -log10() transformed adjusted p-values per gene. 


## Central Volcano Plot


```{r}

genes_to_label <- c("mcl1", "mmp9.1")

central_data <- res$central %>%
  as_tibble(rownames = "gene_id") %>%
  drop_na(padj) %>% # drop all genes with NAs
  filter(padj < 0.5) %>% # reduce the number of points that need to be plotted
  mutate(
    category = case_when(
      padj < 0.05 & log2FoldChange > 0 ~ "Upregulated",
      padj < 0.05 & log2FoldChange < 0 ~ "Downregulated",
      TRUE ~ "Non-significant"
    )
  ) %>%
  left_join(xtrop, by = c("gene_id")) # add annotations

# Calculate the number of upregulated and downregulated genes
n_upregulated <- sum(central_data$category == "Upregulated")
n_downregulated <- sum(central_data$category == "Downregulated")

# Create the plot
gg_central <- central_data %>%
  ggplot(aes(
    x = log2FoldChange, y = -log10(padj), color = category,
    text = paste0("</br>Pcu23 gene: ", gene_id,
                  "</br>X.tr peptide: ", xenp_gene_symbol,
                  "</br>X.tr description: ", xenp_description)
  )) +
  geom_point(alpha = 0.75, shape = 16) +
  scale_color_manual(values = c(
    "Non-significant" = "#808080",
    "Upregulated" = "#FF5733",
    "Downregulated" = "#0F8CBA"
  )) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "#555555") +
  geom_vline(xintercept = 0, linetype = "dashed", color = "#555555") +
  annotate("text", x = 5, y = max(-log10(central_data$padj), na.rm = TRUE) - 0.5, 
           label = paste0("Upregulated: ", n_upregulated), color = "#FF5733", hjust = 1) +
  annotate("text", x = -5, y = max(-log10(central_data$padj), na.rm = TRUE) - 0.5, 
           label = paste0("Downregulated: ", n_downregulated), color = "#0F8CBA", hjust = 0) +
  scale_x_continuous(breaks = seq(-6, 6, by = 2), limits = c(-6, 6)) +
  theme_minimal() +
  theme(
    legend.position = "none",
  ) +
  geom_text_repel(
    data = subset(central_data, xenp_gene_symbol %in% genes_to_label),
    aes(label = xenp_gene_symbol),
    size = 3,
    box.padding = unit(1.7, "lines"),
    point.padding = unit(0.1, "lines"),
    color = "black",
    segment.color = 'black'
  )

gg_central

ggsave("central_volcanoplot.jpeg", width = 7, height = 5 , dpi = 600)
```


## South Volcano Plot

```{r}

genes_to_label <- c("mmp11.1", "ass1", "sec14l3", "stard8", "klf9", "hsp90aa1.1")
extra_gene <- "PECUL23A051069T1"
extra_gene_2 <- "leprotl1"


south_data <- res$south %>%
  as_tibble(rownames = "gene_id") %>%
  drop_na(padj) %>% # drop all genes with NAs
  filter(padj < 0.5) %>% # reduce the number of points that need to be plotted
  mutate(
    category = case_when(
      padj < 0.05 & log2FoldChange > 0 ~ "Upregulated",
      padj < 0.05 & log2FoldChange < 0 ~ "Downregulated",
      TRUE ~ "Non-significant"
    )
  ) %>%
  left_join(xtrop, by = c("gene_id")) # add annotations

# Calculate the number of upregulated and downregulated genes
n_upregulated <- sum(south_data$category == "Upregulated")
n_downregulated <- sum(south_data$category == "Downregulated")

# Create the plot
gg_south <- south_data %>%
  ggplot(aes(
    x = log2FoldChange, y = -log10(padj), color = category,
    text = paste0("</br>Pcu23 gene: ", gene_id,
                  "</br>X.tr peptide: ", xenp_pep_id,
                  "</br>X.tr description: ", xenp_description)
  )) +
  geom_point(alpha = 0.75, shape = 16) +
  scale_color_manual(values = c(
    "Non-significant" = "#808080",
    "Upregulated" = "#FF5733",
    "Downregulated" = "#0F8CBA"
  )) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "#555555") +
  geom_vline(xintercept = 0, linetype = "dashed", color = "#555555") +
  annotate("text", x = 5, y = max(-log10(south_data$padj), na.rm = TRUE) - 0.5, 
           label = paste0("Upregulated: ", n_upregulated), color = "#FF5733", hjust = 1) +
  annotate("text", x = -5, y = max(-log10(south_data$padj), na.rm = TRUE) - 0.5, 
           label = paste0("Downregulated: ", n_downregulated), color = "#0F8CBA", hjust = 0) +
  scale_x_continuous(breaks = seq(-6, 6, by = 2), limits = c(-6, 6)) +
  theme_minimal() +
  theme(
    legend.position = "none",
  ) +
  geom_text_repel(
    data = subset(south_data, xenp_gene_symbol %in% genes_to_label),
    aes(label = xenp_gene_symbol),
    size = 3,
    box.padding = unit(3, "lines"),
    point.padding = unit(0, "lines"),
    color = "black",
    segment.color = "black",
    max.overlaps = Inf
  ) +
  geom_text_repel(
    data = subset(south_data, transcript_id %in% extra_gene),
    aes(label = xenp_gene_symbol),
    size = 3,
    box.padding = unit(2.7, "lines"),
    point.padding = unit(0, "lines"),
    color = "black",
    segment.color = "black",
    max.overlaps = Inf
  ) +
  geom_text_repel(
    data = subset(south_data, xenp_gene_symbol %in% extra_gene_2),
    aes(label = xenp_gene_symbol),
    size = 3,
    box.padding = unit(3, "lines"),
    point.padding = unit(0, "lines"),
    color = "black",
    segment.color = "black",
    max.overlaps = Inf
  )

gg_south
ggsave("south_volcanoplot.jpeg", width = 7, height = 5 , dpi = 600)

```



## Getting information about interesting genes


We filter in the DESeq2 results object those DEGs which are significantly DE.


```{r}
pull_genes<-function(x, alpha=0.05, lfc=0, signed="both", feature="xenp_gene_symbol") {
  
  # filter by adjusted p
  x<-x %>%
    filter(padj<alpha)
  
  if(signed=="up"){
    x<-x %>%
      filter(log2FoldChange>lfc) %>%
      pull(feature)
  }
  
  if(signed=="down"){
    x<-x %>%
      filter(log2FoldChange<(lfc*-1)) %>%
      pull(feature) 
  }
  
  if(signed=="both"){
    x<-x %>%
      filter(abs(log2FoldChange)>lfc) %>%
      pull(feature) 
  }
  
  ## deal with multiple gene annotations for the same gene (different transcripts)
  
  x<-strsplit(x, ";") %>% unlist() %>% unique()
  x<-x[!is.na(x)]
  
  return(x[!is.na(x)])
}

# now extract all

degs_ids<-lapply(res, FUN=function(x) 
  x %>%
  as_tibble(rownames = "gene_id") %>%
  left_join(xtrop) %>%
    pull_genes(feature = "gene_id", alpha = 0.05, lfc=0, signed = "both")
)
str(degs_ids)
```


After that, we join the annotations to those DEGs and arrange them by their lowest adjusted p-value to get more infromation on those DEGs.


```{r}
extract_info <- function(sig_genes, deseq_result, xtrop) {
  # Convert DESeq results to tibble
  deseq_tibble <- as_tibble(deseq_result, rownames = "gene_id")
  
  # Filter DESeq results for significant DEGs based on gene_id
  filtered_deseq_result <- deseq_tibble %>%
    filter(gene_id %in% sig_genes)
  
  # Perform left join with annotation data
  detailed_info <- filtered_deseq_result %>%
    left_join(xtrop, by = "gene_id")
  
  return(detailed_info)
}

```


```{r}
sig_degs_central <- extract_info(degs_ids$central, res$central, xtrop) %>%
    arrange(padj)
sig_degs_south <- extract_info(degs_ids$south, res$south, xtrop) %>%
    arrange(padj)

intersection <- extract_info(intersect(sig_degs_set$Central, sig_degs_set$South), res$south, xtrop)

write.csv(sig_degs_central, file = "sig_degs_central.csv")
write.csv(sig_degs_south, file = "sig_degs_south.csv")
```



### Thyroid related genes


```{r}
extract_info_by_keywords <- function(deseq_result, keywords, xtrop) {
  as_tibble(deseq_result, rownames = "gene_id") %>%
    left_join(xtrop, by = "gene_id") %>% # Add peptide descriptions
    filter(if_any(everything(), ~ str_detect(., paste(keywords, collapse = "|")))) %>% # Filter based on keywords in any column
    arrange(padj)
}

central_info <- extract_info_by_keywords(res$central, c("thyroid", "duox2"), xtrop)
south_info <- extract_info_by_keywords(res$south, c("thyroid", "duox2"), xtrop)

```









```{r}
# Load necessary libraries
library(dplyr)
library(stringr)

# Keywords of interest related to thyroid hormone

# Function to extract relevant information based on description keywords

extract_info_by_keywords <- function(deseq_result, keywords, xtrop) {
  # Define the columns to process
  character_columns <- c("gene_id", "transcript_id", "peptide_id", "xenx_pep_id", "xenx_gene_symbol", "xenx_description", "xenp_pep_id", "xenp_gene_symbol", "xenp_description")
  as_tibble(deseq_result, rownames = "gene_id") %>%
    left_join(xtrop, by = "gene_id") %>%
    rowwise() %>%
    filter(any(sapply(unlist(strsplit(c_across(all_of(character_columns)), ";")), function(x) any(str_detect(x, paste(keywords, collapse = "|")))))) %>%
    arrange(padj) %>%
    ungroup()
}

# Extract information for central and south based on keywords
central_info <- extract_info_by_keywords(res$central, c("thyroid", "duox2"), xtrop)
south_info <- extract_info_by_keywords(res$south, c("thyroid", "duox2"), xtrop)

# ~ str_detect(., ...): This is the anonymous function created using ~
# str_detect(., ...): Checks if the current column contains any of the specified keywords


# Possibility of combining the information
# combined_info <- full_join(central_info, south_info, by = c("gene_id", "xenp_description"), suffix = c("_central", "_south"))

# Print the combined information
# print(combined_info)
```


#### Central region

```{r}
central_info
```

In the central region, there are no terms associated with the thyroid hormone that are significant (adjusted p-value lower than 0.05)


#### Southern region

```{r}
south_info
```

However for the southern region, we do recover some thyroid related genes that have lower adjusted p-value than 0.05.

* thibz, thyroid hormone induced bZip protein [Source:Xenbase;Acc:XB-GENE-484679]; padj=0.001235922

(the next gene is already over padj 0.05: trip4,	thyroid hormone receptor interactor 4 [Source:Xenbase;Acc:XB-GENE-974186]; padj=0.066677105)



### Urea cycle related genes

```{r}
extract_info_by_keywords(res$central, c("urea", "ass1", "arg2"), xtrop)

extract_info_by_keywords(res$south, c("urea", "ass1", "arg2"), xtrop)
```
No significant DEGs in central.

South:
* ass1, argininosuccinate synthase 1 [Source:Xenbase;Acc:XB-GENE-977784], padj=0.003197872
* arg2, arginase 2 [Source:Xenbase;Acc:XB-GENE-482624], padj=0.003584074


### CORT related genes


```{r}
extract_info_by_keywords(res$central, c("CORT", "corticosterone", "hsp90", "klf9"), xtrop)

extract_info_by_keywords(res$south, c("CORT", "corticosterone", "hsp90", "klf9"), xtrop)
```

No significant DEGs in the central region.

South:
* hsp90aa1.1 heat shock protein 90kDa alpha family class A member 1 [Source:Xenbase;Acc:XB-GENE-49530...] padj=0.005406117
* klf9 Kruppel-like factor 9 [Source:Xenbase;Acc:XB-GENE-483889]; padj=0.034931505


### Leptin/lipid metabolism related genes


```{r}

extract_info_by_keywords(res$central, c("lipid", "leptin", "leprotl1"), xtrop)

extract_info_by_keywords(res$south, c("lipid", "leptin", "leprotl1"), xtrop)

```


### Apoptosis


```{r}

extract_info_by_keywords(res$central, c("mmp11", "apoptosis"), xtrop)

extract_info_by_keywords(res$south, c("mmp11", "apoptosis"), xtrop)

```



#### Interactive Volcano Plots

```{r}
# we can now turn this into an interactive plot:
ggplotly(gg_central, tooltip="text")
ggplotly(gg_south, tooltip="text")
```
