---
title: "DESeq2"
author: "clagrabel"
date: "2024-02-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Running DESeq2

```{r}

#load(file = "C:/Users/tester/Desktop/MADOBIS/TFM/TFM_Github/DESeq2.RData")

pacman::p_load(tidyverse, DESeq2)

dds <- DESeqDataSetFromTximport(
  txi = txi_filtered, 
  colData = exp_design,
  design = ~ waterLevel * region)

```

```{r}

dds
assayNames(dds)

```

dds is a  S4 object system, with slots that can be accessed through @ instead of $ ($ being used for S3 objects).
The core of the object are assays (a single or multiple matrices) with samples as columns and features (genes in our case) as rows. 

Let's check the different assays stored in the dds object.

* counts: stores the counts matrix. If we look at the counts matrix, we can appreciate that the decimal counts have been rounded to integers.

Rows are genes and columns are samples. The dimensions are 32531 genes and 48 samples.


```{r}
#library(tidyverse)

assay(dds, "counts") %>% head()

counts_deseq<-assay(dds, "counts")
dim(counts_deseq)

# There are no NAs in the counts matrix 
sum(is.na(counts_deseq))

```

* avgTxLength: contains the average transcript length, in case it is needed to transform counts estimates into FPKM or TPM which require transcript length.


```{r}

assay(dds, "avgTxLength") %>% head()

```


# Pre-filtering

Reducing the size of the matrix speeds up downstream calculations and reduces memory requirements.

We will only keep rows with counts summing up to 10 or more.

((Does it make sense to do it for a minimal sample number of 4. This is meant to ensure that a gene shows a minimal level of expression across what might be considered a biologically coherent group, which in our case would be for example the 4 replicates per sample?????

In the documentation it says sth like that


```{r}
length(dds[rowSums(counts(dds)) >= 10,])

dds <- dds[rowSums(counts(dds)) >= 10,]
```

We are keeping 19317 genes out of the initial 32531.


# DESeq2 behind the scenes

DESeq2 does the following:

* Estimation of Size Factors:

  + The first step involves controlling for differences in library sizes across the sequencing experiments. Library size can vary due to differences in sequencing depth or sample preparation. If not corrected, these differences can influence the interpretation of gene expression changes. DESeq2 estimates size factors using a median ratio method, where the count for each gene in a sample is divided by a reference value (typically, the median of the ratios of counts across all samples). These size factors are then used to normalize the counts, ensuring that comparisons across samples are fair and not biased by differences in sequencing depth.

* Estimation of Dispersion:

  + After normalization, DESeq2 estimates dispersion for each gene, which measures the variability of counts beyond what would be expected from a Poisson distribution (the simplest model for count data, assuming the mean equals the variance). RNA-seq data often exhibit over-dispersion, where the variance exceeds the mean, partly due to biological variability among replicates. DESeq2 uses a model that assumes counts are over-dispersed and estimates dispersion parameters for each gene. This is crucial for accurate differential expression analysis, as it ensures that the statistical tests account for the inherent variability in the data.

* Negative Binomial GLM Fitting:

  + With size factors and dispersion estimates in hand, DESeq2 fits a generalized linear model (GLM) using a negative binomial distribution for each gene. The negative binomial distribution is suitable for count data with over-dispersion because it introduces an additional parameter (over the Poisson distribution) to model the variance separately from the mean. 


* Wald Statistics Calculation and Multiple Testing Correction:

  + Once the GLM is fitted, DESeq2 uses Wald statistics to test the null hypothesis that there is no difference in expression for each gene between conditions. This involves comparing the estimated coefficients from the GLM against zero. Since thousands of genes are tested simultaneously, there's a high risk of false positives due to multiple testing. To address this, DESeq2 applies corrections for multiple testing, such as the Benjamini-Hochberg procedure, which controls the false discovery rate (FDR). This step ensures that the list of differentially expressed genes reported by DESeq2 has a controlled rate of false positives, making the results more reliable and interpretable.


```{r}
dds <- DESeq(dds)
dds
```

_Note: it looks like we are "overwriting" the dds object, but we are actually just adding more information to it_


```{r}
res<- results(dds)
resultsNames(dds)
```


Per gene (row), there are 6 metrics:
* baseMean: Mean of normalized counts for all samples divided by size factor (note: it does not take into account gene length).
* log2FoldChange: The effect size estimate. This value indicates how much the gene or transcript's expression seems to have changed between the comparison and control groups. This value is reported on a logarithmic scale to base 2.
* lfcSE: The standard error estimate for the log2 fold change estimate
* stat: The value of the test statistic for the gene or transcript.
* pvalue: P-value of the test for the gene or transcript.
* padj: Adjusted P-value for multiple testing for the gene or transcript using BH (Benjamini-Hochberg).


### NA values in the Adjusted P-value Column

The padj with NA values are genes which have been filtered out by the independent filtering step in the DESeq2 analysis for having a low mean normalized count.

Let's check some of those counts with NA padj:

```{r}
counts(dds) %>%
  as.data.frame() %>%
  filter(is.na(res$padj)) %>%
  head()
```

Although some there are many 0 values, the sum of the counts per row/gene across the samples is higher than 10 (this is the value that we established before), however their mean count is very low and they are not reliable.


Let's obtain the activated and repressed genes for every comparison.


# Water Level: low vs high, independent of region

```{r}
res_waterLevel <- results(dds, contrast=c("waterLevel", "low", "high"))

summary(res_waterLevel, alpha=0.05)
```

According to the results, there would be 52 genes which are up-regulated in low water level versus high water level, and 7 which are down-regulated. These would be genes that are differentially expressed in both regions only because of the treatment effect.


```{r eval=FALSE, include=FALSE}
genes_waterLevel <- rownames(res_waterLevel)

log_fold_change <- res_waterLevel$log2FoldChange
q_value <- res_waterLevel$padj
names(log_fold_change) <- genes_waterLevel
names(q_value) <- genes_waterLevel

activated_genes_water <- genes_waterLevel[log_fold_change > 1.5 & q_value < 0.05] 
length(activated_genes_water)
length(activated_genes_water[!is.na(activated_genes_water)])
activated_genes_water <- activated_genes_water[!is.na(activated_genes_water)]


repressed_genes_water <- genes_waterLevel[log_fold_change < -1.5 & q_value < 0.05] 
length(repressed_genes_water)
length(repressed_genes_water[!is.na(repressed_genes_water)])
repressed_genes_water <- repressed_genes_water[!is.na(repressed_genes_water)]

```


# Region: Southern Spain vs Central Spain, independent of water level

```{r}
res_region<-results(dds, contrast=c("region", "Southern_Spain", "Central_Spain"))

summary(res_region, alpha=0.05)
```

2154 genes are up-regulated, and 2420 genes  are down-regulated. These would be genes that account for regional differences between populations from Southern Spain and from Central Spain.


```{r eval=FALSE, include=FALSE}

genes_region <- rownames(res_region)
head(genes_region)
log_fold_change <- res_region$log2FoldChange
q_value <- res_region$padj
names(log_fold_change) <- genes_region
names(q_value) <- genes_region

activated_genes_region <- genes_region[log_fold_change > 1.5 & q_value < 0.05]
length(activated_genes_region)
length(activated_genes_region[!is.na(activated_genes_region)])
activated_genes_region <- activated_genes_region[!is.na(activated_genes_region)]

repressed_genes_region <- genes_region[log_fold_change < - 1.5 & q_value < 0.05]
length(repressed_genes_region)
length(repressed_genes_region[!is.na(repressed_genes_region)])
repressed_genes_region <- repressed_genes_region[!is.na(repressed_genes_region)]

#results(dds, contrast=c("region", "Southern_Spain", "Central_Spain"))
```



# Interaction term: interaction between water level and region

```{r}
res_interaction <- results(dds, name="waterLevellow.regionSouthern_Spain", alpha = 0.05)

summary(res_interaction)
```

There seem to be only 14 genes that are differently expressed due to the interaction between region and water level, one of them being up-regulated and the rest of them being down-regulated.



# Specific comparisons for region and water level

In order to visualize the all the comparisons between the conditions (waterLevel and regions) without acquiring information about the interaction between the factors, we will make comparisons between factors: 

* Combine the factors of interest into a single factor with all combinations of the original factors
* Change the design to include just this factor


```{r}

group<-factor(paste0(dds$waterLevel, dds$region))
group_design <- cbind(exp_design, group = group)
group_design

colnames(group_design) <- c("sample_id","waterLevel", "region", "group")

#write.table(group_design, file="group_design.csv", sep=",", row.names = F)

dds_group <- DESeqDataSetFromTximport(
  txi = txi_filtered, 
  colData = group_design,
  design = ~ group
)

dds_group <- dds_group[rowSums(counts(dds_group)) >= 10,]
dds_group <- DESeq(dds_group)

```

Let's check the comparisons:

```{r}
resultsNames(dds_group)
```

## High water in Southern Spain region vs Central Spain

```{r}
res_hS_hC<-results(dds_group, contrast=c("group", "highSouthern_Spain", "highCentral_Spain"))

summary(res_hS_hC, alpha=0.05)
```

2154 up-regulated and 2420 genes down regulated. These would be DEGs accounting for the regional variation in the transcriptome in optimal water conditions, when there is no desiccation stress. 


## Low water in Southern Spain region vs Central Spain

```{r}
res_lS_lC<-results(dds_group, contrast=c("group", "lowSouthern_Spain", "lowCentral_Spain"))

summary(res_lS_lC, alpha=0.05)
```

2474 genes up-regulated and 2736 genes down-regulated. These genes are differentially expressed in the distinct regions in low water conditions (drought or dessication stress). 


```{r eval=FALSE, include=FALSE}
genes_lS_lC <- rownames(res_lS_lC)
head(genes_lS_lC)
log_fold_change <- res_lS_lC$log2FoldChange
q_value <- res_lS_lC$padj
names(log_fold_change) <- genes_lS_lC
names(q_value) <- genes_lS_lC

activated_genes_lS_lC <- genes_lS_lC[log_fold_change > 1.5 & q_value < 0.05]
length(activated_genes_lS_lC)
length(activated_genes_lS_lC[!is.na(activated_genes_lS_lC)])
activated_genes_lS_lC <- activated_genes_lS_lC[!is.na(activated_genes_lS_lC)]

repressed_genes_lS_lC <- genes_lS_lC[log_fold_change < - 1.5 & q_value < 0.05]
length(repressed_genes_lS_lC)
length(repressed_genes_lS_lC[!is.na(repressed_genes_lS_lC)])
repressed_genes_lS_lC <- repressed_genes_lS_lC[!is.na(repressed_genes_lS_lC)]
```


## Southern Spain region high and low water

```{r}
res_lS_hS<-results(dds_group, contrast=c("group", "lowSouthern_Spain", "highSouthern_Spain"))

summary(res_lS_hS, alpha=0.05)
```

#DOUBT
I thought the summary function was already taking out the NAs, but it appears not.
Two ways of getting the actual number of DEGs, excluding 

However, we will extract the DEGs afterwards and we will already exclude the NAs, it's just that I can't trust these numbers that I get from the summary.

```{r}

head(res_lS_hS)

res_lS_hS_filtered_NA<- res_lS_hS %>%
as_tibble() %>%
filter(!is.na(padj))


length(res_lS_hS[is.na(res_lS_hS$padj), ])

res_filtered_south <- res_lS_hS[!is.na(res_lS_hS$padj), ]
summary(res_filtered_south, alpha=0.05)

sum(res_lS_hS$padj < 0.05, na.rm=TRUE)
```



## Central Spain region high and low water


```{r}
res_lC_hC<-results(dds_group, contrast=c("group", "lowCentral_Spain", "highCentral_Spain"))

summary(res_lC_hC, alpha=0.05)
```


## MA Plots

Plotting the mean of normalized counts against the log fold change.

```{r}
par(mfrow=c(3,2))
# single plot
DESeq2::plotMA(res_waterLevel, main="Water Level")
DESeq2::plotMA(res_region, main="Region")
DESeq2::plotMA(res_interaction, main="Interaction")
DESeq2::plotMA(res_hS_hC, main="High Southern vs. Central")
DESeq2::plotMA(res_lS_lC, main="Low Southern vs. Central")
DESeq2::plotMA(res_lS_hS, main="Southern High vs. Low")
DESeq2::plotMA(res_lC_hC, main="Central High vs. Low")
```


# Specific comparisons for Populations and Water Level

We will check for the comparisons for each population in the different water level conditions.

```{r}

group_populations<-factor(paste0(exp_design$waterLevel, exp_design$populations))
group_design_pop <- cbind(exp_design, group = group_populations)
group_design_pop

dds_populations <- DESeqDataSetFromTximport(
  txi = txi_filtered, 
  colData = group_design_pop,
  design = ~ group)

dds_populations <- dds_populations[rowSums(counts(dds_populations)) >= 10,]
dds_populations <- DESeq(dds_populations)
resultsNames(dds_populations)
```

### Buitrago Population: High vs Low Water Level

```{r}
res_hBui_lBui<-results(dds_populations, contrast=c("group", "highBuitrago", "lowBuitrago"))

summary(res_hBui_lBui, alpha=0.05)
```

10 genes up-regulated, 14 genes down-regulated.


### Canencia Population: High vs Low Water Level

```{r}
res_hCan_lCan<-results(dds_populations, contrast=c("group", "highCanencia", "lowCanencia"))

summary(res_hCan_lCan, alpha=0.05)
```

14 genes up-regulated, 38 genes down-regulated.


### Turrubuelo Population: High vs Low Water Level

```{r}
res_hTur_lTur<-results(dds_populations, contrast=c("group", "highTurrubuelo", "lowTurrubuelo"))

summary(res_hTur_lTur, alpha=0.05)
```

4 genes up-regulated, 6 genes down-regulated.


### Espajosas Population: High vs Low Water Level

```{r}
res_hEsp_lEsp<-results(dds_populations, contrast=c("group", "highEspajosas", "lowEspajosas"))

summary(res_hEsp_lEsp, alpha=0.05)
```

14 genes up-regulated, 15 genes down-regulated.


### Llano Population: High vs Low Water Level

```{r}
res_hLla_lLla<-results(dds_populations, contrast=c("group", "highLlano", "lowLlano"))

summary(res_hLla_lLla, alpha=0.05)
```

75 genes up-regulated, 86 genes down-regulated.


### Jabata Population: High vs Low Water Level

```{r}
res_hJab_lJab<-results(dds_populations, contrast=c("group", "highJabata", "lowJabata"))

summary(res_hJab_lJab, alpha=0.05)
```

10 genes up-regulated, 7 genes down-regulated.


## MA Plots

Plotting the mean of normalized counts against the log fold change.

```{r}
par(mfrow=c(1,2))
# single plot
DESeq2::plotMA(res_hBui_lBui, main="Buitrago Population High vs. Low")
DESeq2::plotMA(res_hCan_lCan, main="Canencia Population High vs. Low")
DESeq2::plotMA(res_hTur_lTur, main="Turrubuelo Population High vs. Low")
DESeq2::plotMA(res_hEsp_lEsp, main="Espajosas Population High vs. Low")
DESeq2::plotMA(res_hLla_lLla, main="Llano Population High vs. Low")
DESeq2::plotMA(res_hJab_lJab, main="Jabata Population High vs. Low")
```



# Diagnostics

## P-value distributions
The over-dispersion correction step from DESeq2 should result in an even distribution of p-values, with a peak of genes with low p-values (i.e. a rectangular plot with a peak near 0).


```{r}
# For specific comparisons:
par(mfrow=c(2,2))
par(mar=c(4,4,4,4))
hist(res_hS_hC$pvalue, breaks=50, col="grey", main="hS_hC", xlab = "p-value")
hist(res_lS_lC$pvalue, breaks=50, col="grey", main="lS_lC", xlab = "p-value")
hist(res_lS_hS$pvalue, breaks=50, col="grey", main="lS_hS", xlab = "p-value")
hist(res_lC_hC$pvalue, breaks=50, col="grey", main="lC_hC", xlab = "p-value")

```

```{r}
# For water level and region independently
par(mfrow=c(2,3))
hist(res_waterLevel$pvalue, breaks=50, col="grey", main="Water Level", xlab = "p-value")
hist(res_region$pvalue, breaks=50, col="grey", main="Region", xlab = "p-value")
hist(res_interaction$pvalue, breaks=50, col="grey", main="Interaction", xlab = "p-value")
```

Comparisons that have higher amount of DEGs show a very high peak of p-values of 0. 
However some comparisons are closer to evenly distributed, with the frequency of p-values of 0 being similar to the rest of frequencies, even showing a bit of up-hill distribution meaning that there are more genes with a very high p-value than genes with a low p-value. These comparisons have a much lower amount of DEGs.
I do not think it is necessary to remove outliers or trying to fit another model with more than one dispersion parameter.



## Dispersion plots

To accurately model sequencing counts, we need to generate accurate estimates of within-group variation (variation between biological replicates of the same treatment/condition group) for each gene. With only a few replicates per group, the estimates of variation for each gene are often unreliable. This is because genes, even if they have similar average expression levels, can vary quite a lot in their variability or dispersion. Small sample sizes mean that the estimates of dispersion (variability) can be influenced by random fluctuations in the data, making them less reliable.

To address this problem, DESeq2 shares information across genes within a certain condition to generate more accurate estimates of variation based on the mean expression level of the gene using a method called ‘shrinkage’. DESeq2 assumes that genes with similar expression levels have similar dispersion.

To model the dispersion based on expression level (mean counts of replicates), the dispersion for each gene separately is estimated using maximum likelihood estimation. In other words, given the count values of the replicates, the most likely estimate of dispersion is calculated.

Let's plot these dispersion estimates.

```{r}
plotDispEsts(dds)
plotDispEsts(dds_populations)
```


* The black dots are the starting point of the analysis for each gene. Each dot shows the gene's expression level (how many times it was counted) and its initially estimated dispersion between samples (how much its expression varies) before any adjustments are made.

* The red dots tat form a line, are the expected dispersion of each gene, given how many times it was counted, that is estimated by the model.

* The blue points are the same genes, but 'shrunk' towards the values predicted by the model (the red line), this means after adjusting their dispersion estimates to be closer to what the model predicts. 

* Genes that were not shrunk are shown as black dots outlined in blue.


In general, the plot outcome is good. Out of this plot, we can conclude that:
* The correction model is working, since the black dots (initial estimates) are further away from the red line than the blue dots (adjusted estimates).
* Genes with low counts show a higher dispersion, hence the curve. However, this curve flattens towards genes with higher counts.
* Genes that have much higher dispersion than expected are identified as outliers (black dot with blue circles). These are NOT "shrunk". This is due to the likelihood that the gene does not follow the modeling assumptions and has higher variability than others for biological or technical reasons. Shrinking these could result in the detection of false positives.
* The genes with really low dispersion could be housekeeping genes that are constitutively expressed among most samples.




# Exporting Data


```{r}
# Make a results folder
dir.create("results", showWarnings = FALSE)

# Exporting them all as a list object and save it as an .rds file
saveRDS(list(
          waterLevel = res_waterLevel,
          region = res_region,
          interaction = res_interaction,
          lS_lC = res_lS_lC,
          hS_hC = res_hS_hC,
          lS_hS = res_lS_hS,
          lC_hC = res_lC_hC,
          hBui_lBui = res_hBui_lBui,
          hCan_lCan = res_hCan_lCan,
          hTur_lTur = res_hTur_lTur,
          hEsp_lEsp = res_hEsp_lEsp,
          hLla_lLla = res_hLla_lLla,
          hJab_lJab = res_hJab_lJab),
          file = "./results/deseq2_results_populations.rds")


# export the DESeq2 object as an .rds files
saveRDS(dds, "./results/deseq2_dds.rds")
```


