---
title: "Removing_Llano"
author: "clagrabel"
date: "2024-03-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


We load and store the data into a variable called txi.

```{r}
txi<-readRDS("salmon_gene_counts.rds")
```

We will load the libraries needed for this script.

```{r}
# use pacman to load libraries
pacman::p_load(tidyverse, DESeq2, pheatmap)
```


# Exploring the data

```{r, include=FALSE}
class(txi)
summary(txi)
lapply(X=txi, FUN=head)
```

The txi object is a list with 4 entries, three matrices (abundance, counts, length) and 1 character vector (countsFromAbundance).

The dimensions of each matrix are 106236 rows as the genes and 48 columns as the sample_id. This means the transcripts have been aggregated into 106236 genes.

```{r}
dim(txi$abundance)
```

Here is a brief description of each element in the txi object: 
* Counts: Estimate of the number of reads mapping to each transcript.
* Abundance: Normalized counts in Transcripts Per Million (TPM). This means that per sample, the total counts add up to 1 million. Raw counts cannot be compared across sample_id because each library may vary slightly in terms of the total number of reads, differences in sequencing bias and difference in transcript lengths. 

```{r}
apply(X=txi$abundance, FUN=sum, MARGIN = 2)
```

* Length: Effective length of the target transcript.
* countsFromAbundance: A character vector indicating whether counts were taken directly from the quantifier (salmon) or whether they have been calculated from the abundances by tximport. Default is no (counts are from salmon).


Moreover, we can appreciate that the two first rows correspond to transcripts that mapped against mithocondrial RNA and non-coding RNA (nuclear ribosomal RNA) decoys introduced in the reference transcriptome. 

We will remove these rows.

_Regarding the non-coding RNA, we can see high counts in the counts matrix (without any normalization). Non-coding RNA molecules are abundant in the cell and are often not fully removed during RNA purification processes. All the transcripts were aggregated into one only "gene" or row called nrRNA, therefore showing a higher length  in the length matrix. Longer transcripts are expected to garner more reads simply because they provide more binding sites for the sequencing reads. Thus, in the abundance matrix their abundance is adjusted downward to account for their length._


# Filtering out mtRNA and nrRNA

We have to filter out the first two rows of every matrix. These two rows correspond to the mitochondrial RNA and nuclear ribosomal RNA that mapped against the decoys introduced in the reference transcriptome.

Moreover, there are genes named with the prefix "PECUL23nc" that are also non-coding. We have to find these genes and remove them too.

```{r, include=FALSE}

txi_filtered <- lapply(txi, function(x) {
  # Check if x is a matrix or a data frame
  if (is.matrix(x) || is.data.frame(x)) {
    # Identify rows whose names contain "PECUL23nc"
    rows_to_remove <- grep("PECUL23nc", rownames(x))
    # Combine the rows to remove: the first two rows and rows with specific names
    all_rows_to_remove <- unique(c(1, 2, rows_to_remove))
    # Remove the specified rows
    return(x[-all_rows_to_remove, ])
  } else {
    # If not, return x unchanged
    return(x)
  }
})

# Now, each matrix in the txi list has had the first two rows removed, all the rows containing non-coding RNA removed, and the countsFromAbundance element was not changed.

lapply(X=txi_filtered, FUN=head)
dim(txi_filtered$abundance)
dim(txi_filtered$counts)
dim(txi_filtered$length)

```


# Re-formatting

Changing the names of the columns for the matrices in the txi object.

```{r, include=FALSE}

# install.packages("Hmisc")
library(Hmisc)

# Creating a vector containing the new names for the samples

samples_by_replicate <- sort(unique(Cs(Bui1H, Bui1H, Bui1L, Bui1L, Bui2H, Bui2H, Bui2L, Bui2L, Bui3H, Bui3H, Bui3L, Bui3L, Bui4H, Bui4H, Bui4L, Bui4L, Can1H, Can1H, Can1L, Can1L, Can2H, Can2H, Can2L, Can2L, Can3H, Can3H, Can3L, Can3L, Can4H, Can4H, Can4L, Can4L, Esp1H, Esp1H, Esp1L, Esp1L, Esp2H, Esp2H, Esp2L, Esp2L, Esp3H, Esp3H, Esp3L, Esp3L, Esp4H, Esp4H, Esp4L, Esp4L, Jab1H, Jab1H, Jab1L, Jab1L, Jab2H, Jab2H, Jab2L, Jab2L, Jab3H, Jab3H, Jab3L, Jab3L, Jab4H, Jab4H, Jab4L, Jab4L, Lla1H, Lla1H, Lla1L, Lla1L, Lla2H, Lla2H, Lla2L, Lla2L, Lla3H, Lla3H, Lla3L, Lla3L, Lla4H, Lla4H, Lla4L, Lla4L, Tur1H, Tur1H, Tur1L, Tur1L, Tur2H, Tur2H, Tur2L, Tur2L, Tur3H, Tur3H, Tur3L, Tur3L, Tur4H, Tur4H, Tur4L, Tur4L)))


# Converting the vector's elements into characters using Cs from the package Hmisc, which embeds the terms in quotation marks, selecting only unique names and sorting alphabetically.


# Checking the resulting names variable
samples_by_replicate
class(samples_by_replicate)
length(samples_by_replicate)

# If the length of the names vector equals the number of columns of the matrices, then:

if(length(samples_by_replicate) == ncol(txi_filtered$counts)) {
  # Replace column names for counts, abundance, and length matrix
  colnames(txi_filtered$counts) <- samples_by_replicate
  colnames(txi_filtered$abundance) <- samples_by_replicate
  colnames(txi_filtered$length) <- samples_by_replicate
} else {
  stop("The length of the samples vector does not match the number of sample_id in the txi object.")
}

#class(txi_filtered)

```

We will group all the replicates for a specific condition together, grouping all replicates for high water level and all replicates for low water in the same population.

```{r}

sample_id<-Cs(Bui1H, Bui2H, Bui3H, Bui4H, Bui1L, Bui2L, Bui3L, Bui4L, Can1H, Can2H, Can3H, Can4H, Can1L, Can2L, Can3L, Can4L, Tur1H, Tur2H, Tur3H, Tur4H, Tur1L, Tur2L, Tur3L, Tur4L, Esp1H, Esp2H, Esp3H, Esp4H, Esp1L, Esp2L, Esp3L, Esp4L, Jab1H, Jab2H, Jab3H, Jab4H, Jab1L, Jab2L, Jab3L, Jab4L, Lla1H, Lla2H, Lla3H, Lla4H, Lla1L, Lla2L, Lla3L, Lla4L)

length(sample_id)


if(length(sample_id) == ncol(txi_filtered$counts)) {
  # Reorder the columns for counts, abundance, and length matrices within the list
  txi_filtered$counts <- txi_filtered$counts[, sample_id, drop = FALSE]
  txi_filtered$abundance <- txi_filtered$abundance[, sample_id, drop = FALSE]
  txi_filtered$length <- txi_filtered$length[, sample_id, drop = FALSE]
} else {
  stop("The length of the sample_id vector does not match the number of columns in the txi_filtered$counts matrix.")
}


#class(txi_filtered)
#lapply(X=txi_filtered, FUN=head)

```



We will remove Llano population.


```{r}
lapply(X=txi_filtered, FUN=head)

txi_test <- lapply(txi_filtered, function(x) {
  # Check if x is a matrix or a data frame
  if (is.matrix(x) || is.data.frame(x)) {
    # We want to remove columns that belong to Llano population.
    cols_to_remove <- grep("Lla", colnames(x))
    # Remove the specified rows
    return(x[, -cols_to_remove])
  } else {
    # If not, return x unchanged
    return(x)
  }
})

dim(txi_test$counts)

# Creating the experimental design

exp_design<-read.csv(file= "exp_design.csv", header=T, sep=",")
llano_rows<-grep("Llano", exp_design$populations)
llano_rows

exp_test <- exp_design %>%
              filter(!row_number() %in% llano_rows)
exp_test
```


Running DESeq2.

```{r}
dds_test <- DESeqDataSetFromTximport(
  txi = txi_test, 
  colData = exp_test,
  design = ~ waterLevel * region)

dds_test <- DESeq(dds_test)
```

Storing the results.

```{r}
res_test<- results(dds_test)
resultsNames(dds_test)
```

# Water Level: low vs high, independent of region

```{r}
test_waterLevel <- results(dds_test, contrast=c("waterLevel", "low", "high"))

summary(test_waterLevel, alpha=0.05)

```

According to the results, there would be 48 genes which are up-regulated in low water level versus high water level, and 5 which are down-regulated.


# Region: Southern Spain vs Central Spain, independent of water level

```{r}
test_region<-results(dds_test, contrast=c("region", "Southern_Spain", "Central_Spain"))

summary(test_region, alpha=0.05)
```

There appear to be 1440 genes up-regulated, 1825 genes down-regulated.

# Interaction term: interaction between water level and region

```{r}
test_interaction <- results(dds_test, name="waterLevellow.regionSouthern_Spain", alpha = 0.05)

summary(test_interaction)
```
1 gene up-regulated, 5 genes down-regulated.


# Specific comparisons for region and water level

In order to visualize the all the comparisons between the conditions (waterLevel and regions) without acquiring information about the interaction between the factors, we will make comparisons between factors: 

* Combine the factors of interest into a single factor with all combinations of the original factors
* Change the design to include just this factor


```{r}

group_test<-factor(paste0(dds_test$waterLevel, dds_test$region))
group_test_design <- cbind(exp_test, group = group_test)
group_test_design

colnames(group_test_design) <- c("sample_id","waterLevel", "region", "population", "group")

#write.table(group_design, file="group_design.csv", sep=",", row.names = F)

dds_test_group <- DESeqDataSetFromTximport(
  txi = txi_test, 
  colData = group_test_design,
  design = ~ group
)

dds_test_group

dds_test_group <- dds_test_group[rowSums(counts(dds_test_group)) >= 10,]
dds_test_group <- DESeq(dds_test_group)

```

Let's check the comparisons:

```{r}
resultsNames(dds_test_group)
```

## High water in Southern Spain region vs Central Spain

```{r}
test_hS_hC<-results(dds_test_group, contrast=c("group", "highSouthern_Spain", "highCentral_Spain"))

summary(test_hS_hC, alpha=0.5)
```

## Low water in Southern Spain region vs Central Spain


```{r}
test_lS_lC<-results(dds_test_group, contrast=c("group", "lowSouthern_Spain", "lowCentral_Spain"))

summary(test_lS_lC, alpha=0.5)
```


```{r eval=FALSE, include=FALSE}
genes_lS_lC <- rownames(test_lS_lC)
head(genes_lS_lC)
log_fold_change <- test_lS_lC$log2FoldChange
q_value <- test_lS_lC$padj
names(log_fold_change) <- genes_lS_lC
names(q_value) <- genes_lS_lC

activated_genes_lS_lC <- genes_lS_lC[log_fold_change > 1.5 & q_value < 0.05]
length(activated_genes_lS_lC)
length(activated_genes_lS_lC[!is.na(activated_genes_lS_lC)])
activated_genes_lS_lC <- activated_genes_lS_lC[!is.na(activated_genes_lS_lC)]

repressed_genes_lS_lC <- genes_lS_lC[log_fold_change < - 1.5 & q_value < 0.05]
length(repressed_genes_lS_lC)
length(repressed_genes_lS_lC[!is.na(repressed_genes_lS_lC)])
repressed_genes_lS_lC <- repressed_genes_lS_lC[!is.na(repressed_genes_lS_lC)]
```


## Southern Spain region high and low water

```{r}
test_lS_hS<-results(dds_test_group, contrast=c("group", "lowSouthern_Spain", "highSouthern_Spain"))

summary(test_lS_hS, alpha=0.5)
```


```{r eval=FALSE, include=FALSE}
#Checking the amount of DEGs that surpass a certain threshold for the log2 foldchange.

genes_lS_hS <- rownames(test_lS_hS)
head(genes_lS_hS)
log_fold_change <- test_lS_hS$log2FoldChange
q_value <- test_lS_hS$padj
names(log_fold_change) <- genes_lS_hS
names(q_value) <- genes_lS_hS
length(!is.na(genes_lS_hS[q_value < 0.05]))
activated_genes_lS_hS <- genes_lS_hS[log_fold_change > 1.5 & q_value < 0.05]
length(activated_genes_lS_hS)
length(activated_genes_lS_hS[!is.na(activated_genes_lS_hS)])
activated_genes_lS_hS <- activated_genes_lS_hS[!is.na(activated_genes_lS_hS)]

repressed_genes_lS_hS <- genes_lS_hS[log_fold_change < - 1.5 & q_value < 0.05]
length(repressed_genes_lS_hS)
length(repressed_genes_lS_hS[!is.na(repressed_genes_lS_hS)])
repressed_genes_lS_hS <- repressed_genes_lS_hS[!is.na(repressed_genes_lS_hS)]

# 5 genes are significantly differentially up-regulated and their change in expression is 2^(1.5) times higher in one condition compared to the other. According to this same parameters, 12 genes are down-regulated.

```



## Central Spain region high and low water


```{r}
test_lC_hC<-results(dds_test_group, contrast=c("group", "lowCentral_Spain", "highCentral_Spain"))

summary(test_lC_hC, alpha=0.5)
```


```{r eval=FALSE, include=FALSE}
# Checking the amount of DEGs that surpass a certain threshold for the log2 foldchange.

genes_lC_hC <- rownames(test_lC_hC)
head(genes_lC_hC)
log_fold_change <- test_lC_hC$log2FoldChange
q_value <- test_lC_hC$padj
names(log_fold_change) <- genes_lC_hC
names(q_value) <- genes_lC_hC

activated_genes_lC_hC <- genes_lC_hC[log_fold_change > 1.5 & q_value < 0.05]
length(activated_genes_lC_hC)
length(activated_genes_lC_hC[!is.na(activated_genes_lC_hC)])
activated_genes_lC_hC <- activated_genes_lC_hC[!is.na(activated_genes_lC_hC)]

repressed_genes_lC_hC <- genes_lC_hC[log_fold_change < - 1.5 & q_value < 0.05]
length(repressed_genes_lC_hC)
length(repressed_genes_lC_hC[!is.na(repressed_genes_lC_hC)])
repressed_genes_lC_hC <- repressed_genes_lC_hC[!is.na(repressed_genes_lC_hC)]

# 20 genes are significantly differentially up-regulated and their change in expression is 2^(1.5) times higher in one condition compared to the other. According to this same parameters, 4 genes are down-regulated.

```

# Specific comparisons for population and water level

We will also visualize the all the comparisons between the waterLevel and populations factors, combining the factors of interest into a single factor with all combinations of the original factors and changing the design to include just this factor.


```{r}

group_test_pop<-factor(paste0(dds_test$waterLevel, dds_test$populations))
group_test_design_pop <- cbind(exp_test, group = group_test_pop)
group_test_design_pop

colnames(group_test_design_pop) <- c("sample_id","waterLevel", "region", "population", "group")

#write.table(group_design, file="group_design.csv", sep=",", row.names = F)

dds_test_group_pop <- DESeqDataSetFromTximport(
  txi = txi_test, 
  colData = group_test_design_pop,
  design = ~ group
)

dds_test_group_pop

dds_test_group_pop <- dds_test_group_pop[rowSums(counts(dds_test_group_pop)) >= 10,]
dds_test_group_pop <- DESeq(dds_test_group_pop)

resultsNames(dds_test_group_pop)

```

### Buitrago Population: High vs Low Water Level

```{r}
test_hBui_lBui<-results(dds_test_group_pop, contrast=c("group", "highBuitrago", "lowBuitrago"))

summary(test_hBui_lBui, alpha=0.5)
```

36 genes up-regulated, 63 down-regulated.


### Canencia Population: High vs Low Water Level

```{r}
test_hCan_lCan<-results(dds_test_group_pop, contrast=c("group", "highCanencia", "lowCanencia"))

summary(res_hCan_lCan, alpha=0.5)
```

14 genes up-regulated, 38 down-regulated.


### Turrubuelo Population: High vs Low Water Level

```{r}
test_hTur_lTur<-results(dds_test_group_pop, contrast=c("group", "highTurrubuelo", "lowTurrubuelo"))

summary(test_hTur_lTur, alpha=0.5)
```

11 genes up-regulated, 37 down-regulated.


### Espajosas Population: High vs Low Water Level

```{r}
test_hEsp_lEsp<-results(dds_test_group_pop, contrast=c("group", "highEspajosas", "lowEspajosas"))

summary(test_hEsp_lEsp, alpha=0.5)
```

287 genes up-regulated, 298 down-regulated.


```{r eval=FALSE, include=FALSE}

# I only ran this chunk to double check that Llano population was not included in the design.

test_hLla_lLla<-results(dds_test_group_pop, contrast=c("group", "highLlano", "lowLlano"))

summary(test_hLla_lLla, alpha=0.5)
```


### Jabata Population: High vs Low Water Level

```{r}
test_hJab_lJab<-results(dds_test_group_pop, contrast=c("group", "highJabata", "lowJabata"))

summary(test_hJab_lJab, alpha=0.5)
```

20 genes up-regulated, 16 genes down-regulated.


## Diagnostics

### P-value distribution 


```{r}
# For specific comparisons:
par(mfrow=c(2,2))
par(mar=c(4,4,4,4))
hist(test_hS_hC$pvalue, breaks=50, col="grey", main="hS_hC", xlab = "p-value")
hist(test_lS_lC$pvalue, breaks=50, col="grey", main="lS_lC", xlab = "p-value")
hist(test_lS_hS$pvalue, breaks=50, col="grey", main="lS_hS", xlab = "p-value")
hist(test_lC_hC$pvalue, breaks=50, col="grey", main="lC_hC", xlab = "p-value")
```


### Dispersion plot

```{r}
plotDispEsts(dds_test)
```



# Exporting the results

```{r}

# Exporting them all as a list object and save it as an .rds file
saveRDS(list(
          test_waterLevel = test_waterLevel,
          test_region = test_region,
          test_interaction = test_interaction,
          test_lS_lC = test_lS_lC,
          test_hS_hC = test_hS_hC,
          test_lS_hS = test_lS_hS,
          test_lC_hC = test_lC_hC),
          file = "./results/deseq2_results_test.rds")


# Export the DESeq2 object as an .rds files
saveRDS(dds_test, "./results/deseq2_dds_test.rds")

```

